connector/src/main/scala/com/datastax/spark/connector/datasource/CassandraInJoinReaderFactory.scala

```diff
@@ -0,0 +1,167 @@
+package com.datastax.spark.connector.datasource
+
+import com.datastax.oss.driver.internal.core.cql.ResultSets
+import com.datastax.spark.connector.cql.{CassandraConnector, TableDef}
+import com.datastax.spark.connector.rdd.ReadConf
+import com.datastax.spark.connector.rdd.reader.PrefetchingResultSetIterator
+import com.datastax.spark.connector.util.Logging
+import com.datastax.spark.connector.writer.{CassandraRowWriter, QueryExecutor}
+import com.datastax.spark.connector.{CassandraRow, CassandraRowMetadata, ColumnName, RowCountRef, SomeColumns}
+import com.google.common.util.concurrent.SettableFuture
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.connector.read.{InputPartition, PartitionReader, PartitionReaderFactory}
+import org.apache.spark.sql.sources.In
+import org.apache.spark.sql.types.{LongType, StructField, StructType}
+
+import scala.concurrent.ExecutionContext
+import scala.util.{Failure, Success}
+
+case class CassandraInJoinReaderFactory(
+  connector: CassandraConnector,
+  tableDef: TableDef,
+  inClauses: Seq[In],
+  readConf: ReadConf,
+  schema: StructType,
+  cqlQueryParts: ScanHelper.CqlQueryParts)
+    extends PartitionReaderFactory{
+
+  override def createReader(partition: InputPartition): PartitionReader[InternalRow] =
+    if (cqlQueryParts.selectedColumnRefs.contains(RowCountRef)) {
+      CassandraInJoinCountReader(connector, tableDef, inClauses, readConf, schema, cqlQueryParts, partition)
+    } else {
+      CassandraInJoinReader(connector, tableDef, inClauses, readConf, schema, cqlQueryParts, partition)
+    }
+}
+
+
+abstract class CassandraBaseInJoinReader(
```

Comment:
CassandraInJoinReader is just a concrete subclass of this class.  Should we perhaps just make the current CassandraBaseInJoinReader into CassandraInJoinReader and make CassandraInJoinCountReader extend that?

Reply:
So the reason this is like that is because you cannot extend case classes, so the only way to keep them case classes is to extend a common class